### ISSUE TITLE
Type errors in state change subscriptions upgrading to 2.0.1

### ISSUE THREAD

**Issue by MichaelTurbe** [DanWahlin/Observable-Store#42]
Hi there, I'm using state services with the includeStateChangesOnSubscribe: true setting, and after updating to the latest version, my project isn't building and all of the subscriptions are showing type errors.  Since I was using includeStateChangesOnSubscribe: true, in each subscription I was using the separate state and stateChanges objects as so:
`this.fooStateServiceSubscription = this.fooStateService.stateChanged
    .subscribe(stateEvent => {
      if (stateEvent.stateChanges.pages) { this.allMyPages = stateEvent.state.pages; }`

The compiler now thinks that the stateEvent object dispatched through the description is the state type itself, not the js object that had state and stateChanges on it.

I could go into all of these spots and cast the stateEvent object to any, but that seems rough. Is it possible to ratchet back down the type safety of the dispatched state event object, or somehow make an interface that would work for both versions?

**Comment by DanWahlin** [comment]
Hi Micheal. Thanks for the info. That definitely sounds like an issue with some of the type updates made in the new version. I’ll have to look into it more. I’m traveling a lot right now but will take a look as soon as I can. In the meantime, I’d revert back to the 1.x version (which I’m guessing you’ve already done).

**Comment by DanWahlin** [comment]
I had some free time on a flight today and considered this more. One of the big requests leading up to V2 was to return a strongly-typed state object so I'd like to keep that in there. I played around with several different union types with generics and nothing works super well there to accommodate returning T and another object for the state with changes. 

After thinking about it more, what if we include an interface that looks something like the following:

```typescript
export interface StateWithChanges<T> {
    state: T,
    stateChanges: Partial<T>
}
```

Then on your store/service where **includeStateChangesOnSubscribe** is used (it can now be set globally as well for the entire store), you would extend ObservableStore this way:

export yourClass extends ObservableStore<StateWithChanges<YourState>> { } 

This approach would still require some minimal code changes for V2, but it would allow standard state or the state with changes to be used with full intellisense. 

Open to feedback here so let me know your thoughts when you can. Love to accommodate people mainly using stateChanged as well as you (and I) using stateChanged with changes.

**Comment by MichaelTurbe** [comment]
Definitely a supporter of the strongly typed state change events, so happy to have a breaking change in v2 that will facilitate both strategies.

**Comment by DanWahlin** [comment]
Another concept I'm thinking through (which I really like so far) is adding a new observable that would return a strongly-typed changes object. Something like these:

stateChanged: T (this is the existing one)
stateChangedWithChanges: StateWithChanges<T> (this would be the new one)

That way you wouldn't even have to define a setting like includeStateChangesOnSubscribe and could subscribe directly to the stateChangedWithChanges observable instead. I'm not sure on the name yet for the new one but open to ideas if you have any there.

**Comment by MichaelTurbe** [comment]
I think that's the ideal solution.

**Comment by DanWahlin** [comment]
I'm currently testing out having the the different observables mentioned above. Any preference on naming? Thinking about:

stateChangedWithChanges

OR

stateChangedIncludeChanges

OR

Any better name that you think fits it?

**Comment by MichaelTurbe** [comment]
stateWithChanges rolls off the tongue, but any of those are great. :)

**Comment by DanWahlin** [comment]
I agree - it's different from stateChanged as well which is probably good. I'll look at changing to that. Naming is hard. :-)

**Comment by DanWahlin** [comment]
I just released 2.1.0 which now has `stateWithPropertyChanges`. The `includeStateChangesOnSubscribe` property is now deprecated buts works the same. I'll keep it in there up through a 3.0 release.  The readme has details about the new observable properties. See the version history for the specifics for 2.1.

Thanks for your feedback!
